### ISSUE TITLE
Returning template type (query bus example)

### ISSUE THREAD

**Issue by unkind** [vimeo/psalm#4431]
Hi,

https://psalm.dev/r/8d1b6ba1c6 — it looks like a bug:

```php
// ...

/**
 * @template R as object
 * @psalm-param Query<R> $q
 * @return R
 */
function query(Query $q): object
{
    // ...
}
```

> The inferred type 'Top10PlayersResult' does not match the declared return type 'R:fn-query as object' for query

`Top10PlayersResult` is object type, I'm not sure what's wrong here.

**Comment by weirdan** [comment]
I think this could be a useful inference rule:
```
Query<R> & Top10Players => Query<R> & Query<Top10PlayersResult> => R is Top10PlayersResult
```
@muglug what do you think?

**Comment by muglug** [comment]
@weirdan that inference rule is in effect in the pasted code.

The false positive is _only_ a bug because `Top10Players` is declared `final` – otherwise unsafe behaviour can ensue (as demonstrated in [this issue](https://github.com/facebook/hhvm/issues/8761) that Hack fails to catch).

Given the finalness is a bit of an edge-case, I may or may not get around to fixing this.

**Comment by muglug** [comment]
Yup, I've decided this isn't really a bug, insofar as the `final` makes it too much of an edge-case.  If you can find a variant that doesn't require `final` classes to be sound I'll consider reopening.
